### DISCUSSION TITLE
Null Get Document Request

### DISCUSSION THREAD

**Discussion by ZiaMeza** [GodotNuts/GodotFirebase#416] [Q&A] [answered]
First, starting off by saying thank you so much for helping run this plugin! I've been floored by how easy it has been to use and how perfect it has been for my game, which is based around learning Korean words.

Now I believe I've reached a deadend I cannot address. I have the Firebase database and there are over 100 accounts registered on it. However, recently, I've been getting an error when trying to load data from one account in particular.

![image](https://github.com/user-attachments/assets/50f11b8e-d631-4a9b-8111-8d44bce0dff7)

I have some print statements that show the following:

(Token hidden)
```
 Loading from cloud.
Attempting to load game data from Firestore for user: 5k1Lk8qu---------------------
Fetching document for user ID: 5k1Lk8qudfQB0CdcJ5bQdDz65Ee2
 [Firebase Error] >> Action in error was: 0 (GET DOCUMENT)
get_document returned null for mastery_data 5k1Lk8qudfQB0CdcJ5bQdDz65Ee2
[Error] Failed to retrieve document for user: 5k1Lk8qu---------------------
```

**In summary, the following four things are true:**

1. Until a few days ago, the account would save and load on the local (executable) version just fine. It stopped recently.

2. I can see the account in the Firebase, and indeed can load and save it perfectly fine when running the HTML5 version of the game on itch.io

3. Every other account works fine when loading on the local version. It's specifically this one account that is causing this null result. 

Might you have any ideas for troubleshooting or what may be occurring? My hunch is because the data structure is a dictionary of dictionaries, it has become ?too large? and so is having trouble being loaded locally? Though I don't know why that would prevent it being loaded in the HTML5 version however.

This is my first time making a post like this on GitHub, so I'm unsure how many more details to provide or how much code may be relevant, but I can provide any relevant snippets as needed. Many thanks for your time.

**Comment by WolfgangSenff** [comment]
Hm! That is very weird. I've never heard of something working in HTML5 and not the rest, doh. I doubt it's the size, although something did just occur to me. We have a timeout set somewhere that I think is like 5 seconds? I'll get a link real quick, one moment.

**Reply by WolfgangSenff** [reply]
Oh, btw, is this Godot 4.x?

**Reply by WolfgangSenff** [reply, marked as answer]
If it's 4.x, try going into your copy of the addon and updating this line to a higher number: https://github.com/GodotNuts/GodotFirebase/blob/fa306a52e2c84980912955dd00b91abe36027723/addons/godot-firebase/firestore/firestore.gd#L232

If it's 3.x, you should be able to find a really similar line in that same file, just search for ".timeout".

**Reply by ZiaMeza** [reply]
Oh apologies! This is Godot 3.5.

In case helpful, here's some relevant code I have going on so far:

```
extends Control
var COLLECTION_ID = "mastery_data"
var save_file_path = "user://save_game.json"

func _on_FirebaseAuth_login_succeeded(auth):
	print("Login succeeded.")
	var user_id = auth.get("localid")
	load_game_from_firestore(user_id)
	Firebase.Auth.save_auth(auth)
	disable_sync_button()
	reset_menu_ui()

func load_game_from_firestore(user_id: String):
	print("Attempting to load game data from Firestore for user: ", user_id)
	var collection: FirestoreCollection = Firebase.Firestore.collection(COLLECTION_ID)
	print("Fetching document for user ID: ", user_id)  # Debugging the fetch process start
	
	var document: FirestoreDocument = yield(collection.get_doc(user_id), "completed")
	if document != null:
		print("Document retrieved successfully for user: ", user_id)
		_handle_document(document)
	else:
		print("[Error] Failed to retrieve document for user: ", user_id)


func _handle_document(document: FirestoreDocument):
	if document != null:
		print("Received document from Firestore. Checking validity...")
	else:
		print("[Error] Received a null document from Firestore.")
		
	if is_instance_valid(document):
		print("Document is valid. Processing data...")
		var doc_data = document.document
		print("Raw document data: ", doc_data)  # Debugging raw data
		var converted_data = firestore_to_godot(doc_data)
		var json_data = JSON.print(converted_data)
		print("Converted data for deserialization: ", json_data)
		if not converted_data.empty():
			deserialize_game_state(json_data)
			disable_sync_button()
			print("Game data deserialized successfully.")
		else:
			print("[Warning] Document exists but contains no data or failed to convert.")
	else:
		print("[Error] Document object is invalid or null.")

```

**Reply by ZiaMeza** [reply]
Hey! Setting the timeout higher did the trick. I set it to 10 seconds as a test and it finally loaded as expected! I'm a bit surprised it takes longer than 5 seconds, so perhaps there's some optimizations I can make to reduce that down. Thanks so much!

**Reply by WolfgangSenff** [reply]
I'm thinking also about just removing that value. It has caused issues for us before, and I don't think a timeout is necessarily required. I'll have to think about how to do it, but I'm glad that fixed it!

**Reply by ZiaMeza** [reply]
Oh not a question but an update. Adjusting the timeout has worked perfectly in allowing my local copy to work, but it has revealed to me that all operations seem to take longer when running the program through an executable locally compared to my HTML version.

On my local executable for example, loading my account with ~1000 nested dictionaries can take about 8 seconds, and saving can take anywhere from 3 to 8 seconds for a single save.

When I run the program through the identical HTML5 version on itch.io, loading the account can take about 2 seconds, and saving can take anywhere from 1 to 3 seconds for a single save.

With my accounts that have smaller datasets, the difference is almost negligible and almost instantaneous.

I'm unsure if this is a fixable thing on my end, nor is it a question, but I hope the information is helpful.

**Reply by WolfgangSenff** [reply]
I'm...not sure! I literally have no clue why that would be the case. I'll keep it in mind at least though, thanks! It does sound more like possibly a Godot thing, not anything specific to our code, but again, not sure.
